.travis.yml

```diff
@@ -16,12 +16,9 @@ python:
   # PyPy:
   - pypy3
 
-env:
-
-
 install:
   - if [[ "$TEST_SOCKETCAN" ]]; then sudo bash test/open_vcan.sh ; fi
-  - travis_retry pip install .[test]
+  - travis_retry pip install .[test,mf4]
```

Comment:
Maybe this is not the right way to configure the tests. Maybe we should simply not support the combination of MF4 & PyPy (as mentioned above). This could be done by only installing `travis_retry pip install .[test]` and instead changing `setup.py` to:

```python
import platform
is_cpython = 'CPython' == platform.python_implementation()

tests_require = [
    'pytest~=4.3',
    'pytest-timeout~=1.3',
    'pytest-cov~=2.6',
    'codecov~=2.0',
    'six',
    'hypothesis'
] + extras_require['serial']

if is_cpython:
    tests_require += extras_require['mf4']
```

A very simple not in the docs should be added as well:
`.. note:: *MF4* is currently only supported on CPython and not on PyPy.`

Reply:
https://www.python.org/dev/peps/pep-0508/#environment-markers

```python
'mf4; platform_python_implementation == "CPython"'
```
Maybe?  Untested, but definitely want to get away from arbitrary code relating to setup.  Also maybe `!= "PyPy"` depending on intent.

Reply:
This does not seem to work on my machine: I get `python-can 3.2.0 does not provide the extra 'mf4'` on CPython 3.6. Removing the spaces did not help either.

Reply:
@danielhrisca Sorry for the long delay. With the above change, this should be ready to finally be merged.

Reply:
@felixdivo, I think maybe I misread the code there.  Sorry.  The point is to document dependencies in a way that can be retrieved regardless of what Python is being used to run setup.py.  For example, what I wrote (or something like it) can go into a wheel intact and be processed when the wheel is installed.  `if is_cpython:` would be processed when the wheel is built and then when the wheel is used it would be fixed according to the Python used to create the wheel.  This is also quite relevant to locking dependencies.

But...  I was thinking at a different layer.  I don't know that there's a tidy all around solution for this yet.  Applying the environment markers to each dependency of the mf4 extra could make it a no-op at least and then it could be unconditionally (code-wise) included wherever.  Maybe that's what I would do, I'm not quite sure.

It would be nice if the extra could be made conditionally available but my naive attempt didn't work.

https://gist.github.com/altendky/357e6a19691912593080f18ddcc39235

It would also be nice if setup.py could reference its own extras as dependencies like `.[mf4]` or such but last I played with that I didn't find anything that worked.

Reply:
I got that idea, but it's largely unsupported for now. :( I was just trying what is implemented and what not. That shortcoming of `setup.py` is really annoying as well. -.-